internal/pkg/instrumentation/probe/sampling/sampling.go

```diff
@@ -0,0 +1,305 @@
+	config      any
+}
+
+func traceIDRationDefaultConfig() traceIDRatioConfig {
+	return traceIDRatioConfig{samplingRateNumerator: samplingRateDenominator}
+}
+
+var (
+	_ encoding.BinaryMarshaler   = &samplerConfig{}
+	_ encoding.BinaryUnmarshaler = &samplerConfig{}
+)
+
+func (sc *samplerConfig) MarshalBinary() ([]byte, error) {
+	buf := make([]byte, 0, sampleConfigSize)
+	writingBuffer := bytes.NewBuffer(buf)
+
+	err := binary.Write(writingBuffer, binary.NativeEndian, sc.samplerType)
+	if err != nil {
+		return nil, err
+	}
+
+	err = binary.Write(writingBuffer, binary.NativeEndian, sc.config)
+	if err != nil {
+		return nil, err
+	}
+
+	if available := writingBuffer.Available(); available > 0 {
+		_, _ = writingBuffer.Write(make([]byte, available))
+	}
+
+	return writingBuffer.Bytes(), nil
+}
+
+func (sc *samplerConfig) UnmarshalBinary(data []byte) error {
+	if len(data) != sampleConfigSize {
+		return fmt.Errorf("invalid data size for sampler config: %d", len(data))
+	}
+	readingBuffer := bytes.NewReader(data)
+
+	err := binary.Read(readingBuffer, binary.NativeEndian, &sc.samplerType)
+	if err != nil {
+		return err
+	}
+
+	switch sc.samplerType {
+	case samplerAlwaysOn, samplerAlwaysOff:
+		return nil
+	case samplerTraceIDRatio:
+		var numerator uint64
+		err := binary.Read(readingBuffer, binary.NativeEndian, &numerator)
+		if err != nil {
+			return err
+		}
+		sc.config = traceIDRatioConfig{numerator}
+	case samplerParentBased:
+		var parentBased parentBasedConfig
+		err := binary.Read(readingBuffer, binary.NativeEndian, &parentBased)
+		if err != nil {
+			return err
+		}
+		sc.config = parentBased
+	}
+
+	return nil
+}
+
+// NewSamplingConfig creates a new SamplingConfig from the given eBPF collection.
+// It applies the sampler configuration from the environment variables.
+func NewSamplingConfig(c *ebpf.Collection) (*Config, error) {
+	samplersConfig, ok := c.Maps[samplersConfigMapName]
+	if !ok {
+		return nil, fmt.Errorf("map %s not found", samplersConfigMapName)
+	}
+
+	probeActiveSampler, ok := c.Maps[probeActiveSamplerMapName]
+	if !ok {
+		return nil, fmt.Errorf("map %s not found", probeActiveSamplerMapName)
+	}
+
+	sc := &Config{
+		samplersConfig:     samplersConfig,
+		probeActiveSampler: probeActiveSampler,
+	}
+
+	err := sc.applySamplerFromEnv()
+	if err != nil {
+		return nil, err
+	}
+
+	return sc, nil
+}
+
+func defaultParentBasedSampler() parentBasedConfig {
+	return parentBasedConfig{
+		Root:             alwaysOnID,
+		RemoteSampled:    alwaysOnID,
+		RemoteNotSampled: alwaysOffID,
+		LocalSampled:     alwaysOnID,
+		LocalNotSampled:  alwaysOffID,
+	}
+}
+
+// taken from go.opentelemetry.io/otel/sdk/trace and adapted to work with eBPF.
+func (sc *Config) applySamplerFromEnv() error {
+	defaultParentBased := defaultParentBasedSampler()
+	defaultTraceIDRatio := samplerConfig{
+		samplerType: samplerTraceIDRatio,
+		config:      traceIDRationDefaultConfig(),
+	}
+
+	sampler, ok := os.LookupEnv(tracesSamplerKey)
```

Comment:
This is really low in the program to be interpreting environment variables. Instead, we should have the instrumentation continue to be the place where these are interpreted and have it pass a configuration down to this level.